compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/inference/InferenceUtils.kt

```diff
@@ -45,6 +47,30 @@ fun ConeKotlinType.isKFunctionType(session: FirSession): Boolean {
             kind == FunctionClassKind.KSuspendFunction
 }
 
+fun ConeClassLikeType.isSubtypeOfFunctionalType(session: FirSession, expectedTypeParameterCount: Int? = null): Boolean {
+    val fullyExpandedType = fullyExpandedType(session)
+    val klass = fullyExpandedType.lookupTag.toSymbol(session)?.fir as? FirClass<*> ?: return false
+    return klass.superTypeRefs.any {
```

Comment:
That's right, thanks! Extended the bb test a little bit to cover that case.

Used `AbstractTypeChecker.isSubtypeOf(...)`, but note that it doesn't handle generic types well, e.g.,

`isSubtypeOf(..., Function1<Int, String>, Function1<T, String>) == false`, which seems to me a subtype relation (since type variable `T` could be `Int`). If I dig all the way down to where it returns `false`, type variance of `T` is `in`, and then it chose to check the subtype of 1st type argument the other way around, like `isSubtypeOf(T, Int)`. If this is an expected behavior, fair enough. It simply means we still need to check if the argument type is a functional type. That also bothers unification below; see my other comments.

Reply:
Sorry, I didn't get the where the problem arises
Actually, now your branch has changed drastically since the comment was applied
Let's synchronize our views on the issue.

We've got an argument `argument`  and its `argumentType`
Also, there's a parameter of `parameterType`

We need to check if suspend conversion is needed (and new `expectedType` if it is needed).
The conversion is needed if and only if 
1. `parameterType` is suspend function type `SuspendFunctionN<Args...>` (not a subtype of it, and not kFunctionType). Then we construct `expectedType` as `FunctionN<Args...>`
2. `argumentType` might be a subtype of that `expectedType` and it might _not_ be a subtype of `SuspendFunctionN<Args...>`

The second condition is a bit complicated to check, but as an approximation we can use check subtyping from `argumentType` to `FunctionN<*, *, *>` (all star projections are used instead of real arguments because they  always fit in supertype position)

Reply:
So, `replaceArgumentsWithStarProjections` is the magic wand. Now checking `isSubtypeOfFunctionalType` looks really simple and intuitive. Thanks!